src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs

```diff
@@ -169,7 +169,7 @@ private class TypeStoreItem : StoreItem
         {
             private readonly object _syncRoot = new object();
             private readonly Type _type;
-            private Dictionary<string, PropertyStoreItem> _propertyStoreItems;
+            private volatile Dictionary<string, PropertyStoreItem> _propertyStoreItems;
```

Comment:
Looks ridiculously amenable to `Lazy<T>`, although I'm not sure of the performance implications.

Reply:
`Lazy` is of course the recommended solution, but the whole point of double-checked locking is when you really really need the performance (at the expense of a nest of potential correctness issues.) I didn't want to change the code too much, and of course, such optimisations should be benchmarked but I don't know if such benchmarks have been written.

Reply:
The current access to this looks like this:
```C#
if (_propertyStoreItems == null)
{
    lock (_syncRoot)
    {
        if (_propertyStoreItems == null)
        {
            _propertyStoreItems = CreatePropertyStoreItems();
        }
    }
}

return _propertyStoreItems.TryGetValue(propertyName, out item);
```
It'd be better if it were instead:
```C#
Dictionary<string, PropertyStoreItem> items = _propertyStoreItems;
if (items == null)
{
    lock (_syncRoot)
    {
        items = _propertyStoreItems ?? (_propertyStoreItems = CreatePropertyStoreItems());
    }
}

return items.TryGetValue(propertyName, out item);
```
as in the common case where it's already initialized, that'll avoid an unnecessary secondary volatile read of a field.